### PR TITLE
Calculate dedicated instance

### LINKED ISSUE #123: NestedVirtualization support
Hello,
for an instance, to help to calculate the price, it will be nice to have the NestedVirtualization flag displayed or managed ..
Can you improve in this way please ?

### PR DESCRIPTION
**Pull request by outscale-vbr**
Calculate dedicated instance and add tag tenancy and nested_virtualization
Closes #123

### CODE REVIEW

File: src/core/vms.rs [hunk 1]
```diff
@@ -48,7 +51,10 @@ impl ResourceTrait for Vm {
 
     fn price_per_hour(&self) -> Result<f32, ResourceError> {
         match self.price_per_hour {
-            Some(price) => Ok(price),
+            Some(price) => match self.nested_virtualization {
```
Comment (outscale-mgo) [inline src/core/vms.rs:54]: why is nested_virtualization require to compute price_dedicated_vm_additional_per_hour ?

Comment (outscale-vbr) [reply src/core/vms.rs]: You are right, it is the contrary.
nested_virtualization require to have dedicated

Comment (outscale-vbr) [reply src/core/vms.rs]: https://docs.outscale.com/fr/userguide/Activer-la-virtualisation-imbriqu%C3%A9e.html

**Review by outscale-mgo — APPROVED** [review]

File: src/core/vms.rs [hunk 2]
```diff
@@ -48,7 +51,10 @@ impl ResourceTrait for Vm {
...
+                true => Ok(price + ((price * self.price_dedicated_vm_additional_per_hour) / 100.0)),
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core/vms.rs]: Yes you are right.
I had the doc

File: src/core/vms.rs [hunk 3]
```diff
@@ -47,8 +50,15 @@ impl ResourceTrait for Vm {
...
 
     fn price_per_hour(&self) -> Result<f32, ResourceError> {
+        // formula to calculate dedicated instance (https://docs.outscale.com/en/userguide/Getting-the-Price-of-Your-Resources.html)
         match self.price_per_hour {
-            Some(price) => Ok(price),
+            Some(price) => match self.tenancy.as_str() {
+                "dedicated" => {
+                    Ok(price + ((price * self.price_dedicated_vm_additional_per_hour) / 100.0))
+                }
+                "default" => Ok(price),
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core/vms.rs:59]: Yes you are right.
I improve it :)

File: docs/README.md [hunk 4]
```diff
@@ -19,6 +19,7 @@ Read license for more details.
   - Volumes
   - Public Ips
   - Snapshots (🚨 Warning: snapshot computation is currently known to be over-priced.)
+  - Dedicated Vm (🚨 Warning: dedicated vm computation can be over-priced depend on the situation (Look at example of Pricing for Dedicated VMs [Outscale API](https://docs.outscale.com/api) .)
```
Comment (outscale-mdr) [inline docs/README.md:22]: "overpriced"  and "depending"

Comment (outscale-mdr) [reply docs/README.md:22]: The suitable link is this one, no ? https://docs.outscale.com/en/userguide/Getting-the-Price-of-Your-Resources.html#_calculating_the_price_of_dedicated_vms

File: src/core/vms.rs [hunk 5]
```diff
@@ -47,8 +50,15 @@ impl ResourceTrait for Vm {
...
+                _ => Ok(price),
```
Comment (outscale-mdr) [inline src/core/vms.rs:60]: This should be an error or a warning, no ?

File: src/oapi/vms.rs [hunk 6]
```diff
@@ -173,6 +173,11 @@ impl Input {
                     continue;
                 }
             };
+            let tenancy = match vm.placement.clone().into_iter().next() {
```
Comment (outscale-mdr) [inline src/oapi/vms.rs:176]:
```rust
match vm.placement.clone().into_iter().next() {

}
```
is too complex, just match on `vm.placement`. Something like that:
```rust
match &vm.placement {
    Some(placement) => placement.tenancy.clone().unwrap_or_default(),
    None => todo!(),
}
```

File: src/core/vms.rs (hunk 3 above)
Comment (outscale-mdr) [inline src/core/vms.rs:59]: This should be done in compute, not in price_per_hour, otherwise, the price_per_month is wrong

**Review by outscale-mdr — CHANGES_REQUESTED** [review]

File: src/core/vms.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core/vms.rs]: Done

File: docs/README.md (hunk 4 above)
Comment (outscale-vbr) [reply docs/README.md]: Done

File: src/core/vms.rs (hunk 5 above)
Comment (outscale-vbr) [reply src/core/vms.rs]: Done

File: src/oapi/vms.rs (hunk 6 above)
Comment (outscale-vbr) [reply src/oapi/vms.rs]: Done

File: src/core/vms.rs (hunk 3 above)
Comment (outscale-vbr) [reply src/core/vms.rs]: Done

File: src/core/vms.rs [hunk 7]
```diff
@@ -41,12 +44,30 @@ impl ResourceTrait for Vm {
         price_per_hour += (self.vm_ram_gb as f32) * self.price_license_per_ram_gb_per_hour;
         price_per_hour += self.price_license_per_vm_per_hour;
         price_per_hour += self.price_box_per_hour;
-        self.price_per_hour = Some(price_per_hour);
+        self.price_per_hour = match Some(price_per_hour) {
```
Comment (jerome-jutteau) [inline src/core/vms.rs:47]: you also want to update through `price_per_hour += ` as `self.price_per_month` is not using `self.price_per_hour` just below

File: src/oapi/vms.rs [hunk 8]
```diff
@@ -379,6 +396,12 @@ impl VmSpecs {
         Some(self)
     }
 
+    fn parse_dedicated_instance_additional_prices(mut self, input: &Input) -> Option<VmSpecs> {
+        self.price_dedicated_vm_additional_per_hour =
```
[the remark this answers is not in the document]
Comment (jerome-jutteau) [reply src/oapi/vms.rs:400]: I agree, you can just compute additions to have normal price THEN apply a multiplication factor as a final step.

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]

File: src/oapi/vms.rs [hunk 9]
```diff
@@ -173,6 +173,15 @@ impl Input {
                     continue;
                 }
             };
+
+            let tenancy = match &vm.placement {
+                Some(placement) => placement.tenancy.clone().unwrap_or_default(),
+                None => {
+                    eprintln!("error to get tenancy");
+                    "".to_string()
```
[the remark this answers is not in the document]
Comment (outscale-mgo) [reply src/oapi/vms.rs:181]: is tenancy a require argument for VM ? if yes, there should be an error, if it's optional, it should be using a default value, and silent the None

File: src/core/vms.rs [hunk 10]
```diff
@@ -41,12 +44,30 @@ impl ResourceTrait for Vm {
...
+                _ => {
+                    eprintln!("error to get tenancy");
+                    Some(price)
+                }
+            },
+            None => {
+                eprintln!("error can not calculate price_per_hour");
+                Some(0.0)
+            }
+        };
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core/vms.rs]: Done

**Comment by outscale-vbr** [comment]
We wait until we will compare the digest we get after one week and the same price with osc-cost with the same vmType.

**Comment by albundy83** [comment]
Hello, any chance you release soon ?

**Comment by outscale-vbr** [comment]
Hi @albundy83,
We are still testing to have a good estimation of nested virtualization.
Once we find it, we will update and merge this PR.
Best regards,
Vincent

File: src/oapi/dedicated_instances.rs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+use crate::{
+    core::{dedicated_instances::DedicatedInstance, Resource, Resources},
+    VERSION,
+};
+
+use super::Input;
+
+impl Input {
+    pub fn fetch_dedicated_instances(&mut self) -> Result<(), Box<dyn error::Error>> {
```
Comment (outscale-mgo) [inline src/oapi/dedicated_instances.rs:13]: why does it need to be mut ?

File: src/oapi/vms.rs [hunk 12]
```diff
@@ -156,7 +156,7 @@ impl Input {
         Ok(())
     }
 
-    pub fn fill_resource_vm(&self, resources: &mut Resources) {
+    pub fn fill_resource_vm(&mut self, resources: &mut Resources) {
```
Comment (outscale-mgo) [inline src/oapi/vms.rs:159]: does self need to be mut here too ?

Comment (outscale-vbr) [reply src/oapi/vms.rs:159]: If not mutable:
```
error[E0594]: cannot assign to `self.use_dedicated_instance`, which is behind a `&` reference
   --> src/oapi/vms.rs:188:21
    |
188 |                     self.use_dedicated_instance = false
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be written
    |
help: consider changing this to be a mutable reference
    |
159 |     pub fn fill_resource_vm(&mut self, resources: &mut Resources) {
```

File: src/oapi/dedicated_instances.rs (hunk 11 above)
Comment (outscale-vbr) [reply src/oapi/dedicated_instances.rs:13]: No i will update

Comment (outscale-vbr) [reply src/oapi/dedicated_instances.rs]: Ty

**Review by outscale-mgo — APPROVED** [review]

File: src/oapi/vms.rs [hunk 13]
```diff
@@ -383,6 +417,20 @@ impl VmSpecs {
...
+        let dedicated_vm_additional =
+            input.catalog_entry("TinaOS-FCU", "DedicatedInstanceSurplus", "RunInstances")?;
+        self.factor_vm_additional_cost = match self.tenancy.as_str() {
+            "dedicated" => 1.0 + ((dedicated_vm_additional * 10000.0) / 100.0),
+            "default" => 1.0,
+            _ => {
+                warn!("error to calculate factor");
+                0.0
+            }
+        };
```
Comment (jerome-jutteau) [inline src/oapi/vms.rs:430]: - If its an error, you should use `error!`
- Also default to 1.0, that's better than having the whole vm price going to zero.
- Please provide a more contextual warning such as:
```
unknown_tenancy => {
    error!("vm additional costs for {} is not supported", unknown_tenancy);
    1.0
}
```

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]
Here are some change requests.
Also: this PR should include dedicated instance group (as a separate commit and resource)

File: src/oapi/vms.rs [hunk 14]
```diff
@@ -177,6 +177,23 @@ impl Input {
                     continue;
                 }
             };
+
+            let tenancy = match &vm.placement {
+                Some(placement) => placement.tenancy.clone().unwrap_or_default(),
+                None => {
+                    warn!("error to get tenancy");
+                    "".to_string()
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/oapi/vms.rs:186]: tenancy is a string .

File: src/core/vms.rs [hunk 15]
```diff
@@ -70,13 +75,15 @@ impl Default for Vm {
...
-            nested_virtualization: Some(false),
+            nested_virtualization: false,
+            tenancy: "default".to_string(),
             price_vcpu_per_hour: 0.0,
             price_ram_gb_per_hour: 0.0,
             price_box_per_hour: 0.0,
             price_license_per_ram_gb_per_hour: 0.0,
             price_license_per_cpu_per_hour: 0.0,
             price_license_per_vm_per_hour: 0.0,
+            factor_vm_additional_cost: 0.0,
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core/vms.rs]: Ok

File: docs/README.md [hunk 16]
```diff
@@ -19,6 +19,9 @@ Read license for more details.
   - Volumes
   - Public Ips
   - Snapshots (🚨 Warning: snapshot computation is currently known to be over-priced.)
+  - Dedicated Vm (🚨 Warning: dedicated vm computation can be overpriced depending on the situation (Look at example of Pricing for Dedicated VMs [Getting the Price of Your Resources](https://docs.outscale.com/en/userguide/Getting-the-Price-of-Your-Resources.html#_calculating_the_price_of_dedicated_vms)
+
+) .)
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply docs/README.md]: Ok

File: src/oapi/dedicated_instances.rs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    core::{dedicated_instances::DedicatedInstance, Resource, Resources},
+    VERSION,
+};
+
+use super::Input;
+
+impl Input {
+    pub fn fetch_dedicated_instances(&self) -> Result<(), Box<dyn error::Error>> {
+        if self.use_dedicated_instance {
+            warn!("info: use dedicated instance")
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/oapi/dedicated_instances.rs]: Ok

File: src/oapi/dedicated_instances.rs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+                read_date_rfc3339: self.fetch_date.map(|date| date.to_rfc3339()),
+                region: self.region.clone(),
+                price_per_hour: Some(price_per_hour),
+                price_per_month: None,
+            };
+            resources
+                .resources
+                .push(Resource::DedicatedInstance(core_resource));
+        } else {
+            warn!("info: use default instance")
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/oapi/dedicated_instances.rs]: Ok

**Pull request merged**